### DISCUSSION TITLE
Example of how to listen to other providers inside asyncnotifier

### DISCUSSION THREAD

**Discussion by MaximeDelhorme** [rrousselGit/riverpod#2390] [Q&A] [answered]
Hi @rrousselGit,

I'm trying to listen to other providers in an asyncProvider.
Ideally I'd like to do something like that:

```
@riverpod
class AsyncExample extends _$AsyncExample {

 @override
  FutureOr<> build(NoreSureWhatKindOfRef ref) async {
    final data1 = ref.watch(provider1);
    final data2 = ref.watch(provider2);
    return _fetchTodo();
  }
}
```

In  #1767 you mentioned one can use ref.listen(...) in the AsyncNotifier.build which I guess is going in the direction of what I'm trying to do. Could you please give a concrete example on how to do that? Especially what should be the type of "NoreSureWhatKindOfRef" ref?

Merci pour ton aide!
Maxime

**Comment by Velkamhell97** [comment, marked as answer]
Hi!

I am not very familiar with the auto generation of providers, but I do the following:

```dart
class AsyncTodosNotifier extends AsyncNotifier<List<Todo>> {
  @override
  Future<List<Todo>> build() async {
    final listener = ref.listen(provider1, (previous, next) {
      /// listener logic
    });

    ref.onDispose(() {
      listener.close();
    });

    return _fetchTodo();
  }
}
```

and the data type of the `ref` is `AsyncNotifierProviderRef<T>` where `T` is the value data type in this example `List<Todo>`.

I hope i have helped you!.

**Reply by MaximeDelhorme** [reply]
Works like a charm. Thanks a lot.

**Reply by AhmedLSayed9** [reply]
You don't need to manually dispose/close the listener as Riverpod is doing that for you.
